### ISSUE TITLE
Header missing when nesting StackNavigator in DrawerNavigator

### ISSUE THREAD

**Issue by BerndWessels** [react-community/react-navigation#165]
I try to have a
- Root screen = `DrawerNavigator`
  - Home screen
  - Notifications screen = `StackNavigator`
    - Notifications list screen
    - Notification edit screen

I expect now that the drawer shows
- Home screen title
- Notifications screen title
and that the Notifications screen has a header showing
- Notification list screen title
and when navigated to Notification edit screen then header shows
- Back button
- Notification edit screen title

But neither the Notifications screen nor the Notifications list or edit screens show a navigation header.

How is that supposed to work?

Basically I want the drawer to navigate between the major sections of my app and each of the major sections then has its own navigation with its own headers.

Can anybody provide me with an example if that is possible please?

**Comment by zakster12** [comment]
I am interested in the same thing.

**Comment by tonyxiao** [comment]
Seeing the same issue. Additionally if I wrap StackNavigator around DrawerNavigator, then header does show up, but when drawer opens it shows up below the header rather than above of it, which makes it visually jarring and not usable.

**Comment by rafaelmathiias** [comment]
Same problem here, needs to appear above the header bar.

**Comment by satya164** [comment]
Can someone post sample code of when header is not visible when nesting a stack navigator inside drawer navigator? Can't think of a reason why it'd happen.

**Comment by rafaelmmathias** [comment]
`const MainScreenNavigator = DrawerNavigator({
    Recent: {
        screen: RecentChatsScreen
    },
    All: {
        screen: AllContactsScreen
    }
});
MainScreenNavigator.navigationOptions = {
    title: 'My Chats'
};

export default SimpleApp = StackNavigator({
    Home: {
        screen: MainScreenNavigator
    },
    Chat: {
        screen: ChatScreen
    }
});`

![captura de tela 2017-02-04 as 02 16 53](https://cloud.githubusercontent.com/assets/21037492/22615637/10efb1d2-ea80-11e6-9459-b7ea1e9cd1e6.png)

**Comment by satya164** [comment]
@rafaelmmathias it's a drawer navigation inside a stack navigation, not stack inside drawer. Also header is visible.

**Comment by rafaelmmathias** [comment]
```
const SimpleApp = StackNavigator({
    Home: {
        screen: RecentChatsScreen
    },
    Chat: {
        screen: ChatScreen
    }
});

export default MainScreenNavigator = DrawerNavigator({

    Recent: {
        screen: SimpleApp //Stack navigator inside drawer, don't show header, see the print below
    },
    All: {
        screen: AllContactsScreen
    }
});
```

![captura de tela 2017-02-04 as 12 33 24](https://cloud.githubusercontent.com/assets/21037492/22619183/3033c0f0-ead6-11e6-9218-ebe73d0cb6fa.png)

The drawer works perfectly, but header is hidden. Thanks.

**Comment by zakster12** [comment]
@rafaelmmathias, the header is visible but when it's embeded in a Drawer it's now visible. Can you show us working example where we have something like that:
[https://i.stack.imgur.com/7Wtzk.png](url). As you can see a lot of us can't understand how to do that.

**Comment by rafaelmmathias** [comment]
@zakster12, my code of example in index.ios.js:

```
import React from 'react';
import {AppRegistry, Text, View, Button,StyleSheet} from 'react-native';
import {StackNavigator, DrawerNavigator, TabNavigator} from 'react-navigation';

class HomeScreen extends React.Component {
    static navigationOptions = {
        title: 'Welcome'
    };
    render() {
        const {navigate} = this.props.navigation;
        return (
            <View>
                <Text>Hello, Chat App!</Text>
                <Button onPress={() => navigate('Chat', {user: 'Lucy'})} title="Chat with Lucy"/>
            </View>
        );
    }
}
class ChatScreen extends React.Component {
    static navigationOptions = {
        // Nav options can be defined as a function of the navigation prop:
        title: ({state}) => `Chat with ${state.params.user}`
    };
    render() {
        // The screen's current route is passed in to `props.navigation.state`:
        const {params} = this.props.navigation.state;
        return (
            <View>
                <Text>Chat with {params.user}</Text>
            </View>
        );
    }
}

class AllContactsScreen extends React.Component {
    render() {
        return (
            <View>
                <Text>List of all contacts</Text>
                <Button onPress={() => this.props.navigation.navigate('Chat', {user: 'Lucy'})} title="Chat with Lucy"/>
            </View>
        )
    }
}
const SimpleApp = StackNavigator({
    Home: {
        screen: HomeScreen
    },
    Chat: {
        screen: ChatScreen
    }
});

const MainScreenNavigator = DrawerNavigator({
    Recent: {
        screen: SimpleApp
    },
    All: {
        screen: AllContactsScreen
    }
});
MainScreenNavigator.navigationOptions = {
    title: 'My Chats'
};


AppRegistry.registerComponent('App', () => MainScreenNavigator);
```

Can you show me the code of your print, seems to work fine, I don't know where are my mistake.

**Comment by rafaelmmathias** [comment]
I don't know how to do this.
My hope is #179.

**Comment by michaelduminy** [comment]
On Android, the following code shows a drawer opened by a left hamburger menu button which contains tabs and some modal screens. Importantly the drawer opens up covering the header.

The structure is: `Stack` (or main, with a hidden header) > `Drawer` > `Stack` (this supplies the header) > `tabs`.

```javascript
const tabNavigator = TabNavigator({
  Screen1: {
    screen: Screen1,
  },
  Screen2: {
    screen: Screen2,
  },
  Screen3: {
    screen: Screen3,
  },
  Screen4: {
    screen: Screen4,
  },
}, {
    initialRouteName: 'Screen1'
  });

const stackNavigator = StackNavigator({
  Home: {
    screen: tabNavigator,
    path: 'Home'
  },
}, {
    initialRouteName: 'Home',
    navigationOptions: {
      header: (navigation, defaultHeader) => ({
        ...defaultHeader,
        title: (<CustomHeader {...navigation} />),
        left: (<CustomMenuButton {...navigation} />),
        right: [(
          <CustomRightButton key='right1' {...navigation}>
        ), (
          <NotificationButtonContainer key='right2' {...navigation} />
        )],
      })
    }
  });

const drawerNavigator = DrawerNavigator({
  Main: { screen: stackNavigator },
}, {
    contentComponent: SideDrawer
  });

const mainNavigator = StackNavigator({
  LoggedIn: { screen: drawerNavigator },
  Notifications: { screen: NotificationsContainer },
  'Modal 1': { screen: Modal1 },
  'Modal 2': { screen: Modal2 },
  'Modal 3': {
    screen: Modal3,
    navigationOptions: {
      title: 'Modal 3 Screen',
      header: (navigation, defaultHeader) => ({
        ...defaultHeader,
        visible: true,
        left: (<CustomBackButton {...{ back: () => EventHandlers.handle('BACK') }} />)
      })
    }
  },
}, {
    navigationOptions: {
      header: (navProps) => ({
        visible: false, // <<<<<<<<<<< NOTE THIS
        title: (<HeaderTitle {...navigation} />),
        left: (
          <CustomBackButton2 {...navigation} />
        ),
        style: styles.navBar,
      }),
    }
  });
```

**Comment by BerndWessels** [comment]
@michaelduminy In your example the drawer opens underneath the header - this seems very odd. Is this intended?

**Comment by BerndWessels** [comment]
So the following works for me (only tested on Android yet).

- The drawer draws out covering the full height of the screen - as I am used to on Android.
- The Notifications screen is a Stack with its own header visible on top.

To achieve this I had to render the `NotificationsNavigator` from a `NotificationsContainer` component rather than having it right away within the `DrawerNavigator`.

This has a site effect which I am unsure about whether this is intended, good or bad: The `DrawerNavigator` and the `NotificationsNavigator` now have separate redux reducers.

I actually like this but I can't tell if this is how it should be.

`react-navigator` really really really really needs better documentation of all these common nesting scenarios in combination with `redux`. Otherwise people will get frustrated and go back to `react-native-router-flux`.

So here is my solution for now:

`root app component`:

```js
/**
 * Import external dependencies.
 */
import React, {Component} from 'react';
import {BackAndroid, View} from 'react-native';
import {DrawerNavigator, addNavigationHelpers} from 'react-navigation';
import {connect} from 'react-redux';

/**
 * Import local dependencies.
 */
import Drawer from '../../drawer/component';
import Home from '../../home/component';
import Login from '../../login/component';
import Notifications from '../../notifications/component';

/**
 * The root navigator.
 * If your app should have a drawer, this is the place for it.
 * If not, just place any other navigator here.
 */
const RootNavigator = DrawerNavigator({
    Home: {
        screen: Home
    },
    Notifications: {
        screen: Notifications
    },
}, {
    //initialRouteName: 'Notifications',
    // Uncomment this if you want to use your own custom drawer component.
    // contentComponent: Drawer
});

/**
 * Export the reducer for the root navigator.
 */
export const rootNavigationReducer = (state, action) => (
    RootNavigator.router.getStateForAction(action, state)
);

/**
 * The applications root component.
 */
class AppRoot extends React.Component {
    // Invoked immediately after a component is mounted. Initialization that requires DOM nodes should go here.
    // If you need to load data from a remote endpoint, this is a good place to instantiate the network request.
    // Setting state in this method will trigger a re-rendering.
    componentDidMount() {
        // Subscribe to the hardware back button press event on TODO Android.
        BackAndroid.addEventListener('hardwareBackPress', () => {
            let {dispatch, rootNavigation} = this.props;
            // Close the drawer if necessary.
            if (rootNavigation.routes[rootNavigation.index].key === 'DrawerOpen') {
                dispatch({type: 'Navigate', routeName: 'DrawerClose'});
                return true;
            }
            return false;
        });
    }

    // The render function should be pure, meaning that it does not modify component state,
    // it returns the same result each time it's invoked, and it does not directly interact with the browser.
    // If you need to interact with the browser, perform your work in componentDidMount()
    // or the other lifecycle methods instead. Keeping render() pure makes components easier to think about.
    // You can also return null or false to indicate that you don't want anything rendered.
    // When returning null or false, ReactDOM.findDOMNode(this) will return null.
    render() {
        const {rootNavigation, dispatch} = this.props;
        // Are we authenticated?
        if (true) {
            // Render the root navigator.
            return (
                <RootNavigator navigation={addNavigationHelpers({
                dispatch: dispatch,
                state: rootNavigation
                })}/>
            );
        } else {
            // Render the login screen.
            return (
                <Login/>
            );
        }
    }
}

/**
 * Provide store properties to the component.
 */
const mapStateToProps = (state) => {
    return {
        rootNavigation: state.rootNavigation
    };
};

/**
 * Provide store actions to the component.
 */
const mapDispatchToProps = (dispatch) => ({
    dispatch: dispatch
});

/**
 * Connect the component to the store and export it.
 */
export default connect(mapStateToProps, mapDispatchToProps)(AppRoot);
```

`home screen`:

```js
/**
 * Import external dependencies.
 */
import React, {Component} from 'react';
import {Button, View, Text} from 'react-native';
import {connect} from 'react-redux';

/**
 * Import local dependencies.
 */
import {pingCreators, isPinging} from '../../redux/reducers/ping';

/**
 * Import styles.
 */
import styles from './styles';

/**
 * The component.
 */
class HomeContainer extends Component {
    render() {
        const {isPinging, ping} = this.props;
        return (
            <View style={styles.container}>
                <Text>Welcome Home</Text>
                <Text>{JSON.stringify(isPinging)}</Text>
                <Button onPress={()=>{ping();}} title="ping"/>
            </View>
        )
    }
}

/**
 * Provide store properties to the component.
 */
const mapStateToProps = (state) => {
    return {
        isPinging: isPinging(state.ping)
    }
};

/**
 * Provide store actions to the component.
 */
const mapDispatchToProps = (dispatch) => ({
    ping: () => dispatch(pingCreators.ping())
});

/**
 * Connect the component to the store and export it.
 */
export default connect(mapStateToProps, mapDispatchToProps)(HomeContainer)
```

`notification screen`:

```js
/**
 * Bernd Wessels (https://github.com/BerndWessels/)
 *
 * Copyright © 2016 Bernd Wessels. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 *
 * @flow
 */

'use strict';

/**
 * Import external dependencies.
 */
import React, {Component} from 'react';
import {Button, View, Text} from 'react-native';
import {StackNavigator} from 'react-navigation';
import {connect} from 'react-redux';

/**
 * Import local dependencies.
 */
import NotificationDetails from '../notificationDetails/component';
import NotificationList from '../notificationList/component';

/**
 * Import styles.
 */
//import styles from './styles';

const NotificationsNavigator = StackNavigator({
    NotificationList: {
        screen: NotificationList
    },
    NotificationDetails: {
        screen: NotificationDetails
    }
}, {
    headerMode: 'screen',
    navigationOptions: {
        title: 'my notifications'
    }
});

/**
 * The component.       //// TODO TODO TODO !!!!!!!!!!!!!! This might need its own reducer !!!???!!!??? !!!!!!!!!!!!!!!!!!!!!!!!!!
 */
class NotificationsContainer extends Component {
    // static navigationOptions = {
    //     // Title may be a simple string:
    //     title: 'notification',
    //     header: {
    //         right: <Button title="Info" onPress={()=>console.log('press')} />,
    //     },
    // };
    render() {
        return (
            <NotificationsNavigator/>
        )
    }
}

/**
 * Provide store properties to the component.
 */
const mapStateToProps = (state) => {
    return {
    }
};

/**
 * Provide store actions to the component.
 */
const mapDispatchToProps = (dispatch) => ({
});

/**
 * Connect the component to the store and export it.
 */
export default connect(mapStateToProps, mapDispatchToProps)(NotificationsContainer)
```

**Comment by michaelduminy** [comment]
@BerndWessels perhaps the code I pasted relies too heavily on custom components, however, I can confirm that the drawer opens on top of the header, such that it covers the entire side of the screen with no header above it.

**Comment by dually8** [comment]
ES6 class-style documentation would be _great_ if at all possible. Thanks @BerndWessels for your hard work!

**Comment by mkonicek** [comment]
This seems to be similar / same as https://github.com/react-community/react-navigation/issues/216

**Comment by mkonicek** [comment]
Here's what the bug looks like in my app:

![drawer_bug](https://cloud.githubusercontent.com/assets/346214/22697194/abb61362-ed48-11e6-9dac-e0fad266cdc0.gif)

**Comment by rafaelmmathias** [comment]
Guys, take a look at #179.

https://github.com/react-community/react-navigation/issues/179#issuecomment-279230085
Maybe help us.

**Comment by coreyphillips** [comment]
The structure that helped me achieve a header outside the drawer and to have the DrawerNavigator open on top of the StackNavigator without displaying a header is as follows:
```
const Stack = {
	FirstView: {
		screen: FirstView
	},
	SecondView: {
		screen: SecondView
	},
	ThirdView: {
		screen: ThirdView
	}
};

const DrawerRoutes = {
	FirstViewStack: {
		name: 'FirstViewStack',
		screen: StackNavigator(Stack, { initialRouteName: 'FirstView' })
	},
	SecondViewStack: {
		name: 'SecondViewStack',
		screen: StackNavigator(Stack, { initialRouteName: 'SecondView' })
	},
	ThirdViewStack: {
		name: 'ThirdViewStack',
		screen: StackNavigator(Stack, { initialRouteName: 'ThirdView' })
	},
};

const RootNavigator =
	StackNavigator({
		Drawer: {
			name: 'Drawer',
			screen: DrawerNavigator(
				DrawerRoutes,
			),
		},
		...Stack
	},
		{
			headerMode: 'none'
		}
	);
```

**Comment by nico1510** [comment]
That's my workaround too but I think it's not very convenient because you end up with a lot of unnecessary (single-screen) StackNavigators...

**Comment by stephenhandley** [comment]
also can probably dry up the above to 
```
const DrawerRoutes = Object.keys(Stack).reduce((routes, name)=> {
  let stack_name = `${name}Stack`;
  routes[stack_name] = {
    name: stack_name,
    screen: StackNavigator(Stack, {initialRouteName: name})
  };
  return routes;
}, {});
```

**Comment by stephenhandley** [comment]
One issue I'm seeing with the above is back buttons aren't shown when the drawer routes are clicked

**Comment by thibbo** [comment]
@stephenhandley I'm facing the same issue. Any update update on this subject?

**Comment by stephenhandley** [comment]
I ended up just using the underlying component DrawerNavigator uses directly in my app render method
https://github.com/react-native-community/react-native-drawer-layout

```
import DrawerLayout from 'react-native-drawer-layout';
//...

const MainNavigator = StackNavigator({
  Home: {
    screen: Screens.Home
  },
  Quaid: {
    screen: Screens.Randy
  }
});

const drawer_view = (
  <ScrollView>
    <View>
      <Text>INSIDE DRAWER</Text>
    </View>
  </ScrollView>
);

//...
<DrawerLayout
  renderNavigationView={() => drawer_view}
  // ...
>
  <MainNavigator/>
</DrawerLayout>
```

working for me on iOS but haven't checked on Android yet.

**Comment by thibbo** [comment]
@stephenhandley How would you navigate toward your different screens through your drawer ?

**Comment by kelset** [comment]
Pinging OP @BerndWessels since this issue is quite old and related to an old version of the lib. If no answer I'll close in 7 days.

**Comment by kelset** [comment]
Hi there @BerndWessels ,

In an effort to cleanup this project and prioritize a bit, since this issue had no follow up since my last comment I'm going to close it.

If you are still having the issue (especially if it's a bug report) please open a new issue that uses the new Issue Template to provide some more details to help us solve it.

**Comment by maluramichael** [comment]
This problem still persists. Can you reopen this issue?

**Comment by brentvatne** [comment]
can someone create a new issue following the new issue template? we now require that an example that reproduces the problem on https://snack.expo.io or github are posted along with the issue. thanks! @maluramichael - perhaps you? ;)

**Comment by coreyphillips** [comment]
In such a case change the headerMode to `headerMode: "screen"` in the RootNavigator. This will enable headers for all components by default. Then for any component you wish to hide the headers for you can add `headerMode: "none"` & `header: null` to them individually like so:

```
const Stack = {
	FirstView: {
		screen: FirstView,
                navigationOptions: {
			headerMode: "none",
			header: null
		}
	},
	SecondView: {
		screen: SecondView
	},
	ThirdView: {
		screen: ThirdView
	}
};
```

Hope this helps. Let me know if you still have questions or if I was unclear.
